### ISSUE TITLE
Passwords Stored Insecurely 

### ISSUE THREAD

**Issue by alexfornuto** [Libresonic/libresonic#69]
Currently Subsonic (and thus LibreSonic) stores user names and passwords in `/var/subsonic/db/subsonic.script` insecurely:

```
INSERT INTO USER VALUES('$USER','enc:$HEXNUMBER',1234574,0,0,FALSE,'$EMAIL')
```
The value `$HEXNUMBER` can be trivially decoded to reveal a user's password. I, for one, am tired of having to qualify to every new user that they shouldn't make their password something they wouldn't want me to be able to see.

**Comment by NHellFire** [comment]
I'd suggest using bcrypt for password storage. http://www.mindrot.org/projects/jBCrypt/
Unfortunately, it looks like it'll be a lot of work to change it.
Main problem is the REST API. One of the login methods uses md5(password.salt) and changing that would break any clients using it (hopefully none use that or they fix it). It also supports taking a plain text password, but the REST API would still need to accept hex passwords for clients that don't send it in plain text. That's not a problem as after decoding it, it can just be checked against the encrypted password as normal.
Last.fm password is also stored as hex, so that'd need changing to use an API token.

I'm not a Java developer, so this is out of my depth. Hopefully this'll help someone.

**Comment by danielreuther** [comment]
It's probably worth to mention that this is only an issue if you're not using LDAP.

**Comment by fxthomas** [comment]
Anything wrong with upgrading regular logins to bcrypt, but generating a separate, random password ("mobile API token" could be a nice name for this) for the current `md5(token, salt)` mechanism used by the REST API? I think this could improve security while keeping the API compatible.

**Comment by EugeneKay** [comment]
I think that adding a separate API Token would be unnecessarily complex, since everything that exists is built around "user / password". Explaining how to deal with a single-use account code would cause a lot of confusion.

In my opinion(and this is just an opinion, guided mostly by laziness) secure password storage is not Libresonic's job. External authentication mechanisms(ie, LDAP) do a much better job than we ever could, so let's not re-implement it. My suggestion is threefold:

1. Update the documentation(for administrators) to strongly suggest use of external authentication
2. Add a notice to the local password change form along the lines of "Your password will NOT be stored securely"
3. Add support for other external authentication mechanisms eventually

**Comment by danielreuther** [comment]
I agree that having the (existent) option to use an LDAP to store the passwords securely might lower the priority of this issue _a little_. But I still believe that there are many legitimate reasons for not wanting to introduce a dependency to another external system just for password hashing -- even if it's just for the sake of keeping the overall complexity low.

Don't get me wrong, I too would prefer an LDAP over having the web app store the passwords any time. But one size doesn't fit all :)

I would personally go with JCE and something like PBKDF2 (https://www.owasp.org/index.php/Hashing_Java)

Can someone explain the md5(token, salt) bit talked about above? How does that work in a scenario where Libresonic doesn't know the password at all (i.e. LDAP)?

**Comment by danielreuther** [comment]
To answer my own question: LDAP of course suffers of the very same problems you have described above: Since Subsonic doesn't have access to the clear text password, it cannot md5 it with the client-submitted random(!) salt. This is a severe design limitation. Alas, there's not much we can do about it apart from asking client developers to not use that form of authentication.

So, to cut a long story short: If we introduce a password encryption for the DB, we should make it configurable, so that users can decide for themselves if they want to compromise security for the sake of staying compatible with a certain client that implements the s=&t= scheme.

Or am I overlooking something here?

**Comment by khers** [comment]
> In my opinion(and this is just an opinion, guided mostly by laziness) secure password storage is not Libresonic's job. External authentication mechanisms(ie, LDAP) do a much better job than we ever could, so let's not re-implement it.

I see where you are coming from, but configuring an LDAP setup is non-trivial but installing Libresonic behind a proxy for TLS is pretty simple.  I always thought of this as a deficiency in Subsonic, but the maintainer never seemed to take security all that seriously.  If you are going to offer the ability to store passwords at all, it should be done well.
